src/compute/src/render/persist_source.rs

```diff
@@ -0,0 +1,209 @@
+{
+    let worker_index = scope.index();
+
+    let async_stream = async_stream::try_stream! {
+        // We are reading only from worker 0. We can split the work of reading from the snapshot to
+        // multiple workers, but someone has to distribute the splits. Also, in the glorious
+        // STORAGE future, we will use multiple persist shards to back a STORAGE collection. Then,
+        // we can read in parallel, by distributing shard reading amongst workers.
+        if worker_index != 0 {
+            log::trace!("We are not worker 0, exiting...");
+            return;
+        }
+
+        let timeout = Duration::from_secs(60);
+
+        let persist_location = Location {
+            consensus_uri: consensus_uri,
+            blob_uri: blob_uri,
+        };
+
+        let (blob, consensus) = persist_location.open(timeout).await?;
+
+        let persist_client =
+            mz_persist_client::Client::new(timeout, blob, consensus).await?;
+
+
+        let (_write, read) =
+            persist_client.open::<Row, Row, Timestamp, Diff>(timeout, shard_id).await?;
+
+            let mut splits = read
+                .snapshot(timeout, as_of.clone(), NonZeroUsize::new(1).unwrap())
+                .await?
+                .expect("invalid usage");
+
+            assert_eq!(splits.len(), 1);
+            let split = splits.pop().unwrap();
+            let mut iter = read
+                .snapshot_iter(timeout, split)
+                .await?
+                .expect("invalid usage");
+
+            loop {
+                let next = iter.poll_next(timeout).await?;
+                if next.is_empty() {
+                    break;
+                }
+                yield ListenEvent::Updates(next);
+            }
+
+            let mut listen = read
+                .listen(timeout, as_of)
+                .await?
+                .expect("invalid usage");
+
+            loop {
+                let next = listen.poll_next(timeout).await?;
+                for event in next {
+                    yield event;
+                }
+            }
+    };
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let (timely_stream, token) =
+        mz_storage::source::util::source(scope, "persist_source".to_string(), move |info| {
+            let activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+
+            // There is a bit of a mismatch: listening on a ReadHandle will give us an Antichain<T>
+            // as a frontier in `Progress` messages while a timely source usually only has a single
+            // `Capability` that it can downgrade. We can work around that by using a
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
```

Comment:
The timely source should not have a `Capability` and should instead have only `CapabilitySet` types. We can fix that.

Reply:
Should we do that before merging? I seem to recall you had a draft PR that did that, right?

Reply:
I think feel free to merge. That work is in @cjubb39 's hands at the moment, and it isn't the most important thing it does. We can clean this up afterwards. Also @guswynn will want it too. :D